### DISCUSSION TITLE
Add support for d2 diagrams

### DISCUSSION THREAD

**Discussion by sagikazarmark** [quarto-dev/quarto-cli#4002] [Feature Requests]
d2 is a fairly new, but IMO very appealing diagram language and solution: https://d2lang.com

Would be nice to add support for it in Quarto.

**Comment by FeralFlora** [comment]
Yeah, this would be quite nice. Is there any input from the Quarto team on this?

**Comment by elipousson** [comment]
I found a relatively new D2 Pandoc filter on GitHub: https://github.com/ram02z/d2-filter

It is a TypeScript file–not Lua though so I'm unsure how to convert it into a Quarto extension.

**Reply by cderv** [reply]
Thanks for sharing this! 

> It is a TypeScript file–not Lua though so I'm unsure how to convert it into a Quarto extension.

There is a quick discussion about that here: https://github.com/quarto-dev/quarto-cli/issues/5392
Quarto can run JSON filter but the path to filter is not resolved like Pandoc is doing. It must be on PATH I believe. 
Also as a windows user, I know there is issue with extension - on windows the JS json filter should be called as `d2-filter.cmd` which obviously would not work on non windows. 

So a JSON filter can be used using `filters:` YAML field, like a Lua filter, so there is path consideration. See discussion mentioned above. 

Using a Quarto extension could work to help with the path problem, to be relative to projects. And another solution could be to run the JSON filter from Lua. 

So following that idea, I did a quick wrapper of the existing JSON filter to run with Lua

https://github.com/cderv/run-d2-filter

No specific extensive testing yet, I just tried to run it on Windows.  You can have a look as example of the Lua filter, or use it directly. 

Obviously it would be good to do a Lua équivalent for **d2** program directly. Should not be hard. I'll probably give it a try when I find time. 

Hope it helps

**Comment by leandro-costa** [comment]
Maybe https://kroki.io/#support?

**Reply by cderv** [reply]
Thanks for sharing! 

This would be a candidate for another diagram extensions probably.

**Comment by rcannood** [comment]
Inspired by the ram02z/d2-filter, I ended up creating a Quarto extension for D2 written completely in Lua: [`data-intuitive/quarto-d2`](https://github.com/data-intuitive/quarto-d2)

You can define the behaviour using [attributes](https://github.com/data-intuitive/quarto-d2#attributes) in a .d2 codeblock, but also with [global options](https://github.com/data-intuitive/quarto-d2#global-options) (because specifying the theme and layout every time can be annoying). [Tooltips and links](https://github.com/data-intuitive/quarto-d2#interactive-diagrams) will also work if you use a specific set of settings.

**Reply by FeralFlora** [reply]
Nice! Since this is a lua filter, does it also work with Pandoc? Just so I know if I can use this for both or not.

**Reply by rcannood** [reply]
Good point! I'm probably missing a few imports that quarto already takes care of for me. I can tell that the code doesn't work as a plain pandoc filter _yet_, but I should be able to get it to work. I started work on this in data-intuitive/quarto-d2#2.
